utils/pattern/src/frontend/mod.rs

```diff
@@ -9,17 +9,57 @@ mod serde;
 use crate::common::*;
 use crate::Error;
 use crate::PatternOrUtf8Error;
+use crate::SinglePlaceholderPattern;
 #[cfg(feature = "alloc")]
 use crate::{Parser, ParserOptions};
 #[cfg(feature = "alloc")]
 use alloc::{borrow::ToOwned, str::FromStr, string::String};
+use core::ops::Deref;
 use core::{
     convert::Infallible,
     fmt::{self, Write},
     marker::PhantomData,
 };
 use writeable::{adapters::TryWriteableInfallibleAsWriteable, PartsWrite, TryWriteable, Writeable};
 
+#[allow(clippy::unwrap_used)]
+#[cfg(feature = "alloc")]
+impl SinglePlaceholderPattern<str> {
+    pub fn encode_store(prefix: &str, suffix: &str) -> alloc::boxed::Box<str> {
+        alloc::format!(
+            "{}{prefix}{suffix}",
+            char::from_u32(prefix.len() as u32 + 1).unwrap()
+        )
+        .into_boxed_str()
+    }
+}
+
+#[allow(clippy::unwrap_used, clippy::indexing_slicing)]
+impl<Store> SinglePlaceholderPattern<Store>
+where
+    Store: Deref<Target = str>,
+{
+    pub fn prefix(&self) -> &str {
+        let mut chars = self.store.chars();
+        let index = chars.next().unwrap() as usize;
+        if let Some(index) = index.checked_sub(1) {
+            &chars.as_str()[..index]
+        } else {
+            chars.as_str()
+        }
```

Comment:
Issue: this code is _not_ panic-proof. There are constructors that are not `unsafe`; you can make a Pattern pointing to an invalid store, and the docs say "unexpected behavior may occur" but not "panics may occur".

Please resolve the two fallible array getters in this function to debug-assert and then return `""`.

Reply:
I know, that's why I asked how to do this properly